packages/design-system/src/components/primitives/dnd/use-drop.ts

```diff
@@ -0,0 +1,293 @@
+      const { pointerCoordinates } = state.current;
+      if (pointerCoordinates === undefined) {
+        return;
+      }
+
+      const parentRect = partialDropTarget.element.getBoundingClientRect();
+
+      const pointerRelativeToParent = {
+        x: pointerCoordinates.x - parentRect.left,
+        y: pointerCoordinates.y - parentRect.top,
+      };
+
+      const childrenRects = getChildrenRectsMemoized(partialDropTarget.element);
+
+      const closestChildIndex =
+        childrenRects.length === 0
+          ? 0
+          : getClosestRectIndex(childrenRects, pointerRelativeToParent);
+      const closestChildRect = childrenRects[closestChildIndex] as
+        | Rect
+        | undefined;
+
+      const childrenOrientation = getLocalChildrenOrientation(
+        partialDropTarget.element,
+        childrenRects,
+        closestChildIndex
+      );
+
+      const indexAdjustment = getIndexAdjustment(
+        pointerRelativeToParent,
+        closestChildRect,
+        childrenOrientation
+      );
+
+      const indexWithinChildren = closestChildIndex + indexAdjustment;
+
+      const current = state.current.dropTarget;
+      if (
+        current === undefined ||
+        current.element !== partialDropTarget.element ||
+        current.indexWithinChildren !== indexWithinChildren ||
+        !isEqualRect(current.rect, parentRect)
+      ) {
+        const toGlobalCoordinates = (placement: Placement | undefined) =>
+          placement && {
+            ...placement,
+            x: placement.x + parentRect.left,
+            y: placement.y + parentRect.top,
+          };
+
+        const neighbourRect = childrenRects[
+          indexAdjustment === 0 ? closestChildIndex - 1 : closestChildIndex + 1
+        ] as Rect | undefined;
+
+        const placement =
+          toGlobalCoordinates(
+            getPlacementBetween(closestChildRect, neighbourRect)
+          ) ||
+          toGlobalCoordinates(
+            getPlacementNextTo(
+              parentRect,
+              closestChildRect,
+              childrenOrientation,
+              indexAdjustment > 0 ? "forward" : "backward"
+            )
+          ) ||
+          getPlacementInside(parentRect, childrenOrientation);
+
+        const dropTarget: DropTarget<Data> = {
+          ...partialDropTarget,
+          rect: parentRect,
+          indexWithinChildren,
+          placement,
+        };
+
+        state.current.dropTarget = dropTarget;
+        latestProps.current.onDropTargetChange(dropTarget);
+      }
+    };
+
+    const detectTarget = () => {
+      const {
+        edgeDistanceThreshold = 3,
+        isDropTarget,
+        swapDropTarget,
+      } = latestProps.current;
+
+      const { pointerCoordinates } = state.current;
+      const root = rootRef.current;
+
+      // @todo: Cache this?
+      // Not expensive by itself, but it may call isDropTarget multiple times.
+      let candidate =
```

Comment:
ideally we would test it with a something that has 1000+ instances and measure the time we spend there at least just simply using console.time() to get a rough picture of what will happen soon